src/flamegraph/color/palette_map.rs

```diff
@@ -1,84 +1,136 @@
-use std::borrow::Cow;
+use crate::flamegraph::color::Color;
 use std::collections::HashMap;
 use std::fs::File;
 use std::fs::OpenOptions;
 use std::io;
-use std::io::BufRead;
-use std::io::BufReader;
-use std::io::Write;
+use std::io::{BufRead, BufReader};
 use std::path::Path;
 use std::str::FromStr;
 
-pub(crate) struct PaletteMap<'a>(HashMap<Cow<'a, str>, (u8, u8, u8)>);
+/// Mapping of the association between a function name and the color used when drawing information
+/// from this function.
+#[derive(Default)]
+pub struct PaletteMap(HashMap<String, Color>);
 
-impl<'a> PaletteMap<'a> {
-    pub fn load(file: &str) -> quick_xml::Result<Self> {
-        let mut map = HashMap::default();
-        let path = Path::new(file);
-
-        // If the file does not exist, it is probably the first call to flamegraph with a consistent
-        // palette: there is nothing to load.
-        if path.exists() {
-            let file = File::open(path).map_err(quick_xml::Error::Io)?;
-            let file = BufReader::new(file);
-
-            for line in file.lines() {
-                let line = line.map_err(quick_xml::Error::Io)?;
-
-                // A line is formatted like this: NAME -> rbg(RED, GREEN, BLUE)
-                let mut words = line.split("->");
+impl PaletteMap {
+    /// Returns the color value corresponding to the given function name.
+    pub fn get(&self, func: &str) -> Option<Color> {
+        self.0.get(func).cloned()
+    }
 
-                let name = match words.next() {
-                    Some(name) => name,
-                    None => return Err(quick_xml::Error::UnexpectedToken(line)),
-                };
+    /// Inserts a function name/color pair in the map.
+    pub fn insert<S: ToString>(&mut self, func: S, color: Color) -> Option<Color> {
+        self.0.insert(func.to_string(), color)
+    }
 
-                let color = match words.next() {
-                    Some(name) => name,
-                    None => return Err(quick_xml::Error::UnexpectedToken(line)),
-                };
+    /// Provides an iterator over the elements of the map.
+    pub fn iter(&self) -> impl Iterator<Item = (&str, Color)> {
+        self.0.iter().map(|(func, color)| (func.as_str(), *color))
+    }
 
-                if words.next().is_some() {
-                    return Err(quick_xml::Error::UnexpectedToken(line));
-                }
+    /// Builds a mapping based on the inputs given by the reader.
+    ///
+    /// The reader should provide name/color pairs as text input, each pair separated by a line
+    /// separator.
+    ///
+    /// Each line should follow the format: NAME->rgb(RED, GREEN, BLUE)
+    /// where NAME is the function name, and RED, GREEN, BLUE integer values between 0 and 255
+    /// included.
+    ///
+    /// This function will return an [`std::io::Error`] if the input is not correctly formatted.
+    pub fn from_stream(reader: &mut dyn io::BufRead) -> io::Result<Self> {
```

Comment:
Let's make this `from_reader` to be more consistent with other crates.

Reply:
[Done](https://github.com/jonhoo/inferno/pull/77/commits/574a80ba07903eb71fc0fded6759e77cee247afc)